src/collections/componentSet.ts

```diff
@@ -370,19 +412,32 @@ export class ComponentSet extends LazyCollection<MetadataComponent> {
 
       // Build maps of destructive components and regular components as they are added
       // as an optimization when building manifests.
-      if (asDeletion) {
-        component.setMarkedForDelete(true);
+      if (deletionType) {
+        component.setMarkedForDelete(deletionType);
         this.logger.debug(`Marking component for delete: ${component.fullName}`);
-        if (!this.destructiveComponents.has(key)) {
-          this.destructiveComponents.set(key, new Map<string, SourceComponent>());
+        const deletions = this.destructiveComponents[deletionType];
+        if (!deletions.has(key)) {
+          deletions.set(key, new Map<string, SourceComponent>());
         }
-        this.destructiveComponents.get(key).set(this.sourceKey(component), component);
+        deletions.get(key).set(this.sourceKey(component), component);
       } else {
         if (!this.manifestComponents.has(key)) {
           this.manifestComponents.set(key, new Map<string, SourceComponent>());
         }
         this.manifestComponents.get(key).set(this.sourceKey(component), component);
       }
+
+      // something could try adding a component meant for deletion improperly, which would be marked as an addition
+      // specifically the ComponentSet.fromManifest with the `resolveSourcePaths` options which calls
+      // ComponentSet.fromSource, and adds everything as an addition
+      if (
```

Comment:
oh, maybe I understand the flow logic now that I read down to here.  Maybe this could be a function that could run on the delete and non-delete path?

Or those paths could check for existence on its counterpart (deletion would delete it from manifestComponents and non-deletion would just not add it if it's already in deletions?)